src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -59,25 +62,25 @@ public class IndexFileUtils
 
     private static final SequentialWriterOption writerOption = defaultWriterOption;
 
+    /**
+     * Remembers checksums of files so we don't have to recompute them from the beginning of the file whenever appending
+     * to a file. Keeps checksums with respective file lengths so we can detect some file changes like truncation
+     * or external appends which could omit this code, so this is still safe.
+     * Index files are append-only (immutable), so length checking is enough.
+     */
+    private static final Cache<String, Guard<FileChecksum>> checksumCache = Caffeine.newBuilder()
```

Comment:
It doesn't look safe to have `FileChecksum` lifecycle across different index builds. In the worst case, index is rebuilt and generates different content but the same length. The cached CRC would not be correct.

Reply:
If it is ever rebuilt, we'd reset the cached checksum.
The only problem could arise if the index was rebuild externally not using this code and happened to have the same length but different content.

Reply:
I added the tail checksum to the FileChecksum which will include the footer checksum.